ocpairprogramming/src/main/java/com/pino/project/ocpairprogramming/java8/ocp/chapter3/generics/EnhancedChainingComparator.java

```diff
@@ -0,0 +1,12 @@
+package com.pino.project.ocpairprogramming.java8.ocp.chapter3.generics;
+
+import java.util.Comparator;
+
+public class EnhancedChainingComparator implements Comparator<Squirrel> {
+	public int compare(Squirrel s1, Squirrel s2) {
+		Comparator<Squirrel> c = Comparator.comparing(s -> s.getSpecies());
```

Comment:
how does this compile? where does the compiler know what "s" type is? s on the right side should include Squirrel type right? Isn't it rather:
`(Squirrel s) -> s.getSpecies()`
?

Reply:
That's because Squirrel type is inferred by the type provided into the declaration Comparator<Squirrel> .

Reply:
nice didn't know thanks :)